Mathlib/RingTheory/Norm/Transitivity.lean

```diff
@@ -0,0 +1,192 @@
+
+lemma polyToMatrix_cornerAddX :
+    f.polyToMatrix (cornerAddX M k k k) = (-f (M k k)).charmatrix := by
+  simp [cornerAddX, Matrix.add_apply, charmatrix,
+    RingHom.polyToMatrix, ← AlgEquiv.symm_toRingEquiv, map_neg]
+
+lemma eval_zero_det_det : eval 0 (f.polyToMatrix (cornerAddX M k).det).det = (f M.det).det := by
+  rw [← coe_evalRingHom, RingHom.map_det, ← RingHom.comp_apply,
+    evalRingHom_mapMatrix_comp_polyToMatrix, f.comp_apply, RingHom.map_det]
+  congr; ext; simp [cornerAddX, diagonal, apply_ite]
+
+lemma eval_zero_compRingEquiv_det :
+    eval 0 (compRingEquiv m n R[X] <| f.polyToMatrix.mapMatrix <| cornerAddX M k).det =
+      (compRingEquiv m n R <| f.mapMatrix M).det := by
+  simp_rw [← coe_evalRingHom, RingHom.map_det, ← RingEquiv.coe_toRingHom, ← RingHom.comp_apply,
+    ← RingHom.comp_assoc, evalRingHom_mapMatrix_comp_compRingEquiv, RingHom.comp_assoc,
+    RingHom.mapMatrix_comp, evalRingHom_mapMatrix_comp_polyToMatrix, ← RingHom.mapMatrix_comp,
+    RingHom.comp_apply]
+  congr; ext; simp [cornerAddX, diagonal, apply_ite]
+
+theorem compRingEquiv_det_mul_pow :
+    ((f.mapMatrix M).compRingEquiv m n R).det * (f (M k k)).det ^ (Fintype.card m - 1) =
+      (f (M k k)).det * ((f.mapMatrix (mulAuxMatBlock M k)).compRingEquiv _ n R).det := by
+  convert Eq.refl ((f.mapMatrix (M * auxMat M k)).compRingEquiv m n R).det
+  · simp_rw [_root_.map_mul, det_mul, ((auxMat_blockTriangular M k).mapMatrix f).compRingEquiv
+      |>.det_fintype, Fintype.prod_Prop, compRingEquiv_toSquareBlock (b := (· ≠ k)),
+      det_reindex_self, f.mapMatrix_toSquareBlock, auxMat_toSquareBlock_true,
+      auxMat_toSquareBlock_false, smul_one_eq_diagonal, ← diagonal_one, f.mapMatrix_diagonal,
+      compRingEquiv_diagonal, det_reindex_self]
+    simp
+  · simp_rw [((mul_auxMat_blockTriangular M k).mapMatrix f).compRingEquiv.det_fintype,
+      Fintype.prod_Prop, compRingEquiv_toSquareBlock (b := (· = k)), det_reindex_self,
+      f.mapMatrix_toSquareBlock, mul_auxMat_toSquareBlock_true, smul_one_eq_diagonal,
+      f.mapMatrix_diagonal, compRingEquiv_diagonal, det_reindex_self]
+    simp; rfl
+
+variable {M f} in
+lemma det_det_aux
+    (ih : ∀ M, (f M.det).det = ((f.mapMatrix M).compRingEquiv {a // (a = k) = False} n R).det) :
+    ((f M.det).det - ((f.mapMatrix M).compRingEquiv m n R).det) *
+      (f (M k k)).det ^ (Fintype.card m - 1) = 0 := by
+  rw [sub_mul, compRingEquiv_det_mul_pow, ← det_pow, ← map_pow, ← det_mul, ← _root_.map_mul,
+    det_mul_corner_pow, _root_.map_mul, det_mul, mulAuxMatBlock, ih, sub_self]
+
+end Algebra.Norm.Transitivity
+
+open Algebra.Norm.Transitivity
+
+/-- The main result in Silvester's paper *Determinants of Block Matrices*: the determinant of
+a block matrix with commuting, equal-sized, square blocks can be computed by taking two
+determinants in a row: first take the determinant over the commutative ring generated by the
+blocks (`S` here), then take the determinant over the base ring. -/
```

Comment:
Does it also work in the opposite order (I think it should, after 2 secs of thought): first take the determinant of each of the blocks, then take the determinant of the resulting matrix?

Reply:
Answered [on Zulip](https://leanprover.zulipchat.com/#narrow/channel/217875-Is-there-code-for-X.3F/topic/Determinant.20under.20restriction.20of.20scalars/near/489171142) but here's an explicit counterexample: let A be the 2x2 identity matrix and let B be the  permutation matrix [[0,1],[1,0]], then A commutes with everything, det A = 1 and det (-B) = det B = -1. Apply Silvester we get that the determinant of [[A,B],[-B,A]] is $\det (A^2-B(-B)) = \det (A^2+B^2) = \det(2I)=4$, while taking the determinant of [[det A, det B], [det (-B), det A]] results in $1^2-(-1)^2=0$.